src/v/s3/client.cc

```diff
@@ -613,7 +613,11 @@ ss::future<> client::put_object(
       std::move(body),
       [this, timeout, header = std::move(header)](
         ss::input_stream<char>& body) mutable {
-          return _client.request(std::move(header.value()), body, timeout)
+          auto make_request = [this, &header, &body, &timeout]() {
```

Comment:
nit: just reading Seastar docs I think this is fine, but I'm curious, does this get simpler if we write `put_object()` as a coroutine and use a `ss::defer`? Does `body` as an r-value play well with coroutines? Feel free to punt since it's a non-trivial change:

Something like

```
auto header = _requestor.make_unsigned_put_object_request(...)
if (!header) {
    throw std::system_error(header.error());
}
ss::defer([&body] { body.close(); });
try {
    auto stream = co_await _client.request(std::move(header.value()), body, timeout);
    auto buf = co_await drain_response_stream(stream);
    auto status = stream->get_headers().result();
    if (status != ok) { ... }
    co_return;
} catch (...) {
    ...
}

Reply:
`body.close()` returns a future so it might not be possible to use with `ss::defer`.

Reply:
Ah, bummer.

Also, when can `make_unsigned_put_object_request` fail? Do we need to `body.close()` in that case?

Reply:
Good point, it is possible for that call to fail particularly when using IAM roles because auth info is set there. Will look into handling closing the stream there.

Reply:
added code to close the stream when returning early

Reply:
Now that we do this explicitly, do we actually need to call `close()` here?

Reply:
Yes, the two close statements are for different code paths so we need both (for the early return and for the other code path)